### DISCUSSION TITLE
Local Storage Support

### DISCUSSION THREAD

**Discussion by cbaker6** [netreconlab/Parse-Swift#69] [Ideas]
## Primitive Storage
ParseSwift has primitive (key/value) storage by default that automatically saves/updates to a devices [Keychain](https://developer.apple.com/documentation/security/certificate_key_and_trust_services/keys/storing_keys_in_the_keychain) and can be [synchronized with users iCloud accounts](https://swiftpackageindex.com/netreconlab/parse-swift/5.7.0/documentation/parseswift/setaccessgroup(_:synchronizeacrossdevices:)). All `.current()` objects are saved to the primitive storage such as: [ParseUser.current()](https://swiftpackageindex.com/netreconlab/parse-swift/main/documentation/parseswift/parseuser/current()), [ParseInstallation.current()](https://swiftpackageindex.com/netreconlab/parse-swift/main/documentation/parseswift/parseinstallation/current()), [ParseVersion.current()](https://swiftpackageindex.com/netreconlab/parse-swift/main/documentation/parseswift/parseversion/current()), [ParseConfig.current()](https://swiftpackageindex.com/netreconlab/parse-swift/main/documentation/parseswift/parseconfig/current()), etc. It has been shown that the [max size of a single one of the aforementioned key/values is ~16MB](https://stackoverflow.com/a/24165635/4639041). Note, the content of a [ParseFile](https://swiftpackageindex.com/netreconlab/parse-swift/main/documentation/parseswift/parsefile) does not count against the max size as the contents of ParseFile's are stored directly to disk. 

## Cache
ParseSwift also cache's queries and objects saved to a parse-server. Details about how to configure the cache can be found in the [documentation](https://swiftpackageindex.com/netreconlab/parse-swift/main/documentation/parseswift/parseconfiguration). Similar to `Primitive Storage`, the content of a `ParseFile` does not count against the cache as the contents of a `ParseFile` is stored directly to disk.

## Examples of using ParseSwift with Local Storage
- [ParseCareKit](https://github.com/netreconlab/ParseCareKit) is a middleware that depends on Parse-Swift and conforms to a protocol which keeps the local CoreData store in sync using [vector clocks](https://en.wikipedia.org/wiki/Vector_clock) with a Parse Server in a SwiftUI based [sample app](https://github.com/netreconlab/CareKitSample-ParseCareKit). It’s a synchronized distributed database. To get some ideas on how to setup your CoreData store for something like this, I recommend looking at Apple's [CareKit](https://github.com/carekit-apple/CareKit) (specifically the [OCKStore](https://github.com/carekit-apple/CareKit/tree/main/CareKitStore/CareKitStore/CoreData))
  - Note that this is different from the other SDKs on purpose to provide more flexibility to the developer. You can see a discussion about the limitations of the current local storage approach in some of the other SDKs [issue 1](https://github.com/parse-community/Parse-Swift/issues/3#issuecomment-657537085) and [issue2](https://github.com/parse-community/Parse-SDK-iOS-OSX/issues/1387#issue-400547016) and why we currently believe a protocol approach would be better

- #40 attempted to add a local store and may have some useful approaches for design, though it doesn't follow the protocol approach needed to be merged into ParseSwift

**Comment by cbaker6** [comment]
@claesjacobsson I answered your question on the Parse-Community discord, but the Parse PMC moderates my answers and blocks many of them. If you have questions about Parse-Swift, I recommend you ask them on this repo. If you want an up-to-date Swift SDK, I recommend you use this one, you can see why here: https://github.com/netreconlab/Parse-Swift/discussions/7. You can update to [4.16.2](https://github.com/netreconlab/Parse-Swift/releases/tag/4.16.2) with no breaking changes (See #70 for more info). ParseSwift 5.x+ will require changes to your apps that depend on Parse-Swift.

**Comment by claesjacobsson** [comment]
Thanks Corey!

Even though I don't know the details behind the two forks of Parse-Swift, I realise it must be frustrating for you with all the hard work you have put into this project. Sorry about that and thanks for all that you have done!

From my own business' perspective I am wondering if I should be worried to rely on Parse-Swift in my apps. From your point of view, where do you see this going?

**Comment by cbaker6** [comment]
>  Sorry about that and thanks for all that you have done!

Appreciate you!

> From my own business' perspective I am wondering if I should be worried to rely on Parse-Swift in my apps. From your point of view, where do you see this going?

IMO, you can rely on Parse-Swift<sup>OG</sup> in this repo as I use it in my apps and plan on maintaining it for the foreseeable future. Since I never relied on anyone from the parse-community for the original Parse-Swift, there's no limitation or info I need from them to improve or add features to the SDK. In addition, I have strong understanding of the server as I'm [#17 on the parse-server Contributors list](https://github.com/parse-community/parse-server/graphs/contributors) (and I still add [bug fixes](https://github.com/parse-community/parse-server/pulls?q=is%3Apr+is%3Aclosed+author%3Acbaker6+) there when necessary), which means I should understand any server-side feature added in the future and translate that to ParseSwift<sup>OG</sup>. Also, ParseSwift<sup>OG</sup> is stable and modern with version `5.0.0+`, so new features should be easy for me to add, and bug fixes should be straightforward.

When it comes to features, the feature comparison tables in #72 demonstrate why I believe ParseSwift<sup>OG</sup> is currently the best of the Parse SDKs and will remain that way. In addition, you can write Cloud-Code with it using [ParseServerSwift](https://github.com/netreconlab/parse-server-swift)

Lastly, I'll note that ParseSwift was highlighted in [issue 560](https://iosdevweekly.com/issues/560#start) of iOS Dev Weekly and even caught the attention of an original Parse Developer on [Twitter](https://twitter.com/nsunimplemented/status/1530233596014755840?s=46&t=52rhbcYlq0pxnlYiMVOp3w). I do not take that type of attention for granted and believe ParseSwift<sup>OG</sup> can provide a better development experience than frameworks like Firebase and the others that are out there, so it will remain in development.
